nextroute/model_sanity_check.go

```diff
@@ -0,0 +1,163 @@
+	// ExecuteMoveDuration is the duration to execute the best move on a
+	// vehicle.
+	ExecuteMoveDuration SanityCheckData `json:"execute_move_duration"`
+	// MovesFailed is the number of moves that failed. A move can fail if the
+	// estimate of a constraint is wrong.
+	MovesFailed int `json:"moves_failed"`
+	// PlanUnitsNoMoveFound is the number of plan units that have no move found.
+	// These moves can not be planned ever, they are over-constrained.
+	PlanUnitsNoMoveFound int `json:"plan_units_no_move_found"`
+	// PlanUnitsUnPlannable is the number of plan units for which moves have
+	// been found but the move can not be planned. This should not happen.
+	PlanUnitsUnPlannable int `json:"plan_units_un_plannable"`
+	// SanityCheckPlanUnits is the sanity check of the plan units.
+	SanityCheckPlanUnits []*SanityCheckPlanUnit `json:"sanity_check_plan_units"`
+}
+
+// SanityCheckPlanUnit is the sanity check of a plan unit.
+type SanityCheckPlanUnit struct {
+	// ID is the ID of the plan unit. The ID of the plan unit is the slice of
+	// ID's of the stops in the plan unit.
+	Stops []string `json:"stops"`
+	// VehiclesHaveMoves is the number of vehicles that have moves for the plan
+	// unit. Only calculated if the depth is medium.
+	VehiclesHaveMoves int `json:"vehicles_have_moves,omitempty"`
+	// UnPlannable is true if the plan unit is un-plannable. A plan unit is
+	// un-plannable if a move is found but the Move.Execute does not result in a
+	// success.
+	UnPlannable bool `json:"un_plannable"`
+	// NoMoveFound is true if no move is found for the plan unit. A plan unit
+	// has no move found if the plan unit is over-constrained.
+	NoMoveFound bool `json:"no_move_found"`
+}
+
+// SanityCheckVehicles is the sanity check of the vehicles.
+type SanityCheckVehicles struct {
+	// Vehicles is the number of vehicles.
+	Vehicles int `json:"vehicles"`
+	// SanityCheckVehicles is the sanity check of the vehicles.
+	SanityCheckVehicles []*SanityCheckVehicle `json:"sanity_check_vehicles"`
+}
+
+// SanityCheckVehicle is the sanity check of a vehicle.
+type SanityCheckVehicle struct {
+	// ID is the ID of the vehicle.
+	ID string `json:"id"`
+	// PlanUnitsHaveMoves is the number of plan units that have moves for the
+	// vehicle. Only calculated if the depth is medium.
+	PlanUnitsHaveMoves int `json:"plan_units_have_moves,omitempty"`
+}
+
+// SanityCheckData is the data of a sanity check property.
+type SanityCheckData struct {
+	Minimum float64 `json:"minimum"`
+	Maximum float64 `json:"maximum"`
+	Average float64 `json:"average"`
+}
+
+// NewModelSanityCheck returns a new ModelSanityCheck.
+//
+// The sanity check is a function that takes a model and returns a
+// ModelSanityCheck. The sanity check is run on the model. The sanity check
+// should return the sanity check output and an error if the sanity check
+// itself fails.
+//
+//	 func someSolver(model nextroute.Model) (runSchema.Output, error) {
```

Comment:
If the sanity check does not take long it would also be advisable to add it prior to the actual run, right? Should we point that out too?